client/utils/src/pubsub.rs

```diff
@@ -0,0 +1,248 @@
+	shared: Arc<Mutex<Shared<M, R>>>,
+}
+
+/// The receiving side of the subscription.
+///
+/// The messages are delivered as items of a `futures::Stream`.
+/// Upon drop this receiver unsubscribes itself from the `Hub`.
+#[derive(Debug)]
+pub struct Receiver<M, R>
+where
+	R: Unsubscribe,
+{
+	rx: TracingUnboundedReceiver<M>,
+
+	shared: Weak<Mutex<Shared<M, R>>>,
+	subs_id: SeqID,
+}
+
+#[derive(Debug)]
+struct Shared<M, R> {
+	id_sequence: crate::id_sequence::IDSequence,
+	registry: R,
+	sinks: HashMap<SeqID, TracingUnboundedSender<M>>,
+}
+
+impl<M, R> AsMut<R> for Shared<M, R> {
+	fn as_mut(&mut self) -> &mut R {
+		&mut self.registry
+	}
+}
+
+impl<M, R> Hub<M, R>
+where
+	R: Unsubscribe,
+{
+	/// Provide mutable access to the registry (for test purposes).
+	pub fn lock_registry<'a>(&'a self) -> impl DerefMut<Target = impl AsMut<R>> + 'a {
+		self.shared.lock()
+	}
+}
+
+impl<M, R> Drop for Receiver<M, R>
+where
+	R: Unsubscribe,
+{
+	fn drop(&mut self) {
+		if let Some(shared) = self.shared.upgrade() {
+			shared.lock().unsubscribe(self.subs_id);
+		}
+	}
+}
+
+impl<M, R> Hub<M, R> {
+	/// Create a new instance of Hub (with default value for the Registry).
+	pub fn new(tracing_key: &'static str) -> Self
+	where
+		R: Default,
+	{
+		Self::new_with_registry(tracing_key, Default::default())
+	}
+
+	/// Create a new instance of Hub over the initialized Registry.
+	pub fn new_with_registry(tracing_key: &'static str, registry: R) -> Self {
+		let shared =
+			Shared { registry, sinks: Default::default(), id_sequence: Default::default() };
+		let shared = Arc::new(Mutex::new(shared));
+		Self { tracing_key, shared }
+	}
+
+	/// Subscribe to this Hub using the `subs_key: K`.
+	///
+	/// A subscription with a key `K` is possible if the Registry implements `Subscribe<K>`.
+	pub fn subscribe<K>(&self, subs_key: K) -> Receiver<M, R>
+	where
+		R: Subscribe<K> + Unsubscribe,
+	{
+		let mut shared = self.shared.lock();
+
+		let subs_id = shared.id_sequence.next_id();
+		let (tx, rx) = crate::mpsc::tracing_unbounded(self.tracing_key);
+		assert!(shared.sinks.insert(subs_id, tx).is_none(), "Used IDSequence to create another ID. Should be unique until u64 is overflowed. Should be unique.");
+		shared.registry.subscribe(subs_key, subs_id);
+
+		Receiver { shared: Arc::downgrade(&self.shared), subs_id, rx }
+	}
+
+	/// Send the message produced with `Trigger`.
+	///
+	/// This is possible if the registry implements `Dispatch<Trigger, Item = M>`.
+	pub fn send<Trigger>(&self, trigger: Trigger)
+	where
+		R: Dispatch<Trigger, Item = M>,
+	{
+		let mut shared = self.shared.lock();
+		let (registry, sinks) = shared.get_mut();
+
+		registry.dispatch(trigger, |subs_id, item| {
+			if let Some(tx) = sinks.get_mut(&subs_id) {
+				if let Err(send_err) = tx.unbounded_send(item) {
+					log::warn!("Sink with SubsID = {} failed to perform unbounded_send: {} ({} as Dispatch<{}, Item = {}>::dispatch(...))", subs_id, send_err, std::any::type_name::<R>(),
```

Comment:
The dispatch is executed with the lock over the registry held.
Meanwhile there is a defined order between underlying-channel-drop and un-subscription (from the receiver's perspective).

Hence the warning: if the execution flow got there — something unforeseen is afoot.

Reply:
@koute in the light of your recent notes on the order between `sinks.remove->registry.unsubscribe` and `registry.subscribe->sinks.insert`, it's reasonable to keep these warnings: the execution would hit here if the registry's `Subscribe<K>` or `Unsubscribe` had panicked.

Reply:
Applying my recent notes should make this case impossible. (:

Reply:
@koute I've implemented the `ReentrantMutex+RefCell` registry sharing and an exotic receiver-possession scheme in `pubsub::tests::panicking_registry::t02`, in this case that piece of code is executed. 
So I'd keep it a warning rather than an assertion.